{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare jobs for tool comparison of real data (IMG/VR4) \n",
    "We'll test the different aligners/search software on the real data from IMG/VR4.  \n",
    "The original workflow used in IMG/VR4 is described in article [DOI: 10.1093/nar/gkac1037](https://doi.org/10.1093/nar/gkac1037) and is available in the [workflow](https://github.com/jgi-microbiome-data-science/crispr-host-prediction).  \n",
    "That workflow uses a pre-filtered CRISPR sapcer set, the IMG/Vr4 virus sequence set, and spacer-protospacer pairs are identified via `blastn -w 8 --dust no` which is roughly equivalent to `blastn -task short` (what we will use here). Another KEY difference is the `--max_target_seqs` parameter, on the IMG/VR4 workflow is set to 1000, and here to some absurd highly (supposedly) unrealistic value (1000000).  \n",
    "Note, we do not compare any filtereing done after the pairing was done (in the original workflow, a `assign_host.py` script preforms, among else, an evalution of the results). For that reason, we also do not benchmark hots assignment tools like spacerphaser, wish or iphop (even though they may be useful or utilize spacer-protospacer pairs).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To simplify the execution of the different tools, we'll use bash scripts generated by `bench.py` script, which we will modify to use the same contigs and spacers files, and discard the hyperfine. Instead, as the sequence set sizes are considerablt larger than the ones used by bnech.py simulations, we will run the different tools as seperate jobs via SLURM. Later (in [`Resource_usage_imgvr4.ipynb`](./Resource_usage_imgvr4.ipynb)), we will use get a general overview of the performance of the tools."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we download the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir imgvr4_data\n",
    "cd imgvr4_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note, we'll use the IMG_VR_2022-09-20_7.1 version of IMG/VR (v4.1, with bugfix in UViG table and protein fasta), and specifically the high-confidence genomes only ( `IMGVR_all_nucleotides-high_confidence.fna`)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "git clone https://github.com/jgi-microbiome-data-science/crispr-host-prediction \n",
    "cd crispr-host-prediction\n",
    "# Download the database\n",
    "mkdir crispr-db\n",
    "cd crispr-db\n",
    "wget https://portal.nersc.gov/cfs/m342/crisprDB/crispr_spacers_filtered_clustered.tsv --quiet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Prepare the database files\n",
    "cut -f1,14 crispr_spacers_filtered_clustered.tsv | sed 1d | sed 's/^/>/' | tr '\\t' '\\n' > spacers.fna\n",
    "cut -f1,7 crispr_spacers_filtered_clustered.tsv > spacer_lineage.tsv\n",
    "seqkit rmdup spacers.fna -s -i -j4 -o rm_duped_spacers.fna\n",
    "seqkit seq -M100 imgvr4_data/crispr-host-prediction/crispr-db/rm_duped_spacers.fna -w0  > tmp\n",
    "mv tmp imgvr4_data/crispr-host-prediction/crispr-db/rm_duped_spacers.fna\n",
    "seqkit fx2tab -nl ./imgvr4_data/crispr-host-prediction/crispr-db/rm_duped_spacers.fna  > ./imgvr4_data/crispr-host-prediction/crispr-db/rm_duped_spacers_name_length.tab\n",
    "# makeblastdb -in spacers.fna -out spacers.fna -dbtype nucl--quiet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ..\n",
    "#login to https://genome.jgi.doe.gov/portal/pages/dynamicOrganismDownload.jsf?organism=IMG_VR\n",
    "seqkit rmdup imgvr4_data/crispr-host-prediction/IMGVR_all_nucleotides-high_confidence.fna -s -i -j4 -o imgvr4_data/crispr-host-prediction/IMGVR_all_nucleotides-high_confidence_nr.fna\n",
    "# wget https://portal.nersc.gov/genomad/__data__/IMGVR_DATA/imgvr4_nucleotide_db.tar.gz --quiet\n",
    "wget https://portal.nersc.gov/genomad/__data__/IMGVR_DATA/IMGVR4_SEQUENCES.fna --quiet\n",
    "# tar -xzf imgvr4_nucleotide_db.tar.gz    \n",
    "# #BLAST viruses to database\n",
    "#blastn -query test_viruses.fna -db crispr-db/spacers.fna -dust no -word_size 8 -max_target_seqs 1000 -outfmt '6 std qlen slen' -num_threads 64 > blastn.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we launch a dummy benchmark to generate the bash scripts, and make sure the tools installed and working."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/clusterfs/jgi/groups/science/homes/uneri/.micromamba/envs/base_env/lib/python3.10/functools.py:889: DataOrientationWarning: Row orientation inferred during DataFrame construction. Explicitly specify the orientation by passing `orient=\"row\"` to silence this warning.\n",
      "  return dispatch(args[0].__class__)(*args, **kw)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Running bowtie1...\n",
      "Benchmark 1: bowtie-build results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/bt1_contigs_index && bowtie --threads 14 -f --all -v 3 -x results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/bt1_contigs_index results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa -S results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//bowtie1_output.sam\n",
      "  Time (mean ± σ):      6.175 s ±  0.224 s    [User: 4.420 s, System: 0.490 s]\n",
      "  Range (min … max):    6.043 s …  6.434 s    3 runs\n",
      " \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " \n",
      "  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Successfully ran tool: bowtie1\n",
      "\n",
      "Running bowtie2...\n",
      "Benchmark 1: bowtie2-build --large-index --threads 14 results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/contigs_bt2indx && bowtie2 --all --xeq --very-sensitive -x results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/contigs_bt2indx -f results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa -S results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//bowtie2_output.sam --threads 14\n",
      "  Time (mean ± σ):      6.298 s ±  0.090 s    [User: 31.202 s, System: 2.964 s]\n",
      "  Range (min … max):    6.207 s …  6.387 s    3 runs\n",
      " \n",
      "\n",
      "Successfully ran tool: bowtie2\n",
      "\n",
      "Running minimap2...\n",
      "Benchmark 1: minimap2 -N 100 --eqx -t 14 -a results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa -o results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//minimap2_output.sam\n",
      "  Time (mean ± σ):     391.4 ms ±  27.8 ms    [User: 280.8 ms, System: 143.0 ms]\n",
      "  Range (min … max):   362.5 ms … 417.8 ms    3 runs\n",
      " \n",
      "\n",
      "Successfully ran tool: minimap2\n",
      "\n",
      "Running bbmap_skimmer...\n",
      "Benchmark 1: bbmapskimmer.sh sam=1.4 maxindel=0 tipsearch=0 midpad=100 in=results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa ref=./results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa outm=results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//bbmap_skimmer_output.sam t=14 minid=0.85 path=results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs/\n",
      "  Time (mean ± σ):      8.084 s ±  0.035 s    [User: 2.420 s, System: 11.291 s]\n",
      "  Range (min … max):    8.045 s …  8.114 s    3 runs\n",
      " \n",
      "\n",
      "Successfully ran tool: bbmap_skimmer\n",
      "\n",
      "Running strobealign...\n",
      "Benchmark 1: strobealign --eqx -k 15 -N 1000 -t 14 results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa -o results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//strobealign_output.sam\n",
      "  Time (mean ± σ):     136.7 ms ±   5.3 ms    [User: 338.8 ms, System: 44.8 ms]\n",
      "  Range (min … max):   132.8 ms … 142.7 ms    3 runs\n",
      " \n",
      "\n",
      "Successfully ran tool: strobealign\n",
      "\n",
      "Running blastn...\n",
      "Benchmark 1: makeblastdb -in results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa -dbtype nucl -out results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//contigs_blastdb && blastn -query results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa -db results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//contigs_blastdb -max_target_seqs 1000000 -out results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//blastn_output.tsv -evalue 1e-5 -num_threads 14 -task blastn-short -outfmt \"6 qaccver saccver nident length mismatch qlen gapopen qstart qend sstart send evalue bitscore\"\n",
      "  Time (mean ± σ):     762.9 ms ±  46.4 ms    [User: 407.2 ms, System: 61.8 ms]\n",
      "  Range (min … max):   709.5 ms … 791.7 ms    3 runs\n",
      " \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " \n",
      "  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Successfully ran tool: blastn\n",
      "\n",
      "Running mmseqs2...\n",
      "Benchmark 1: mkdir -p results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_spacers results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_contigs results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_mmseqs results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_mmseqs_outputs && mmseqs createdb results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_spacers/mmdb && mmseqs createdb results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_contigs/mmdb && mmseqs search results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_spacers/mmdb results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_contigs/mmdb results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_mmseqs_outputs/mmseqs_output results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_mmseqs --min-seq-id 0.85 --min-aln-len 17 --threads 14 -a --search-type 3 -v 1  && mmseqs convertalis results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_spacers/mmdb results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_contigs/mmdb results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_mmseqs_outputs/mmseqs_output results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//mmseqs_output.tsv --format-mode 0 --search-type 4 --format-output query,target,nident,alnlen,mismatch,qlen,gapopen,qstart,qend,tstart,tend,evalue,bits\n",
      "  Time (mean ± σ):     19.407 s ±  0.222 s    [User: 11.956 s, System: 10.890 s]\n",
      "  Range (min … max):   19.190 s … 19.634 s    3 runs\n",
      " \n",
      "\n",
      "Successfully ran tool: mmseqs2\n",
      "\n",
      "Running spacer_containment...\n",
      "Benchmark 1: spacer-containment --n-threads 14 results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa > results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//spacer_containment_output.tsv\n",
      "  Time (mean ± σ):     561.4 ms ±  53.5 ms    [User: 440.4 ms, System: 474.9 ms]\n",
      "  Range (min … max):   451.2 ms … 628.4 ms    10 runs\n",
      " \n",
      "\n",
      "Successfully ran tool: spacer_containment\n",
      "\n",
      "Running mummer4...\n",
      "Benchmark 1: nucmer --maxmatch --nosimplify --batch=10000000 --threads 14 --sam-long=results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//mummer4_output.sam -c 1 results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa\n",
      "  Time (mean ± σ):      1.417 s ±  0.022 s    [User: 1.208 s, System: 0.110 s]\n",
      "  Range (min … max):    1.392 s …  1.436 s    3 runs\n",
      " \n",
      "\n",
      "Successfully ran tool: mummer4\n",
      "\n",
      "Running lexicmap...\n",
      "Benchmark 1: mkdir -p results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_lexicmap_contigs results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_lexicmap_spacers && cp results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_lexicmap_contigs/simulated_contigs.fa && cp results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_lexicmap_spacers/simulated_spacers.fa && lexicmap index -k 15 -m 40000 --seed-max-desert 200 --seed-in-desert-dist 50 -I results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_lexicmap_contigs -O results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_lexicmap.lmi && lexicmap search -d results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_lexicmap.lmi results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//tmp_lexicmap_spacers/simulated_spacers.fa -o results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//lexicmap_output.tsv --threads 14 --align-min-match-len 17 --align-min-match-pident 85 --seed-min-prefix 15 --seed-min-single-prefix 15 --seed-max-dist 100 --seed-max-gap 100 --align-max-gap 20 --align-band 100 --top-n-genomes 0 -a\n",
      "  Time (mean ± σ):      8.913 s ±  0.128 s    [User: 6.330 s, System: 3.486 s]\n",
      "  Range (min … max):    8.785 s …  9.040 s    3 runs\n",
      " \n",
      "\n",
      "Successfully ran tool: lexicmap\n",
      "\n",
      "Running vsearch...\n",
      "Benchmark 1: vsearch --usearch_global results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_spacers.fa --db results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/simulated_data/simulated_contigs.fa --id 0.85 --maxaccepts 0 --strand both --threads 14 --maxhits 0 --maxrejects 0 --minseqlength 17 --maxgaps 0 --strand both --wordlength 7 --maxseqlength 5000000 --userfield query+target+ids+alnlen+mism+ql+opens+qilo+qihi+tilo+tihi+evalue+bits --userout results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//vsearch_output.blast6 --output_no_hits\n",
      "  Time (mean ± σ):      1.093 s ±  0.050 s    [User: 4.642 s, System: 0.627 s]\n",
      "  Range (min … max):    1.037 s …  1.134 s    3 runs\n",
      " \n",
      "\n",
      "Successfully ran tool: vsearch\n",
      "\n",
      "Reading results for bowtie1...\n",
      "File size: 0.02 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//bowtie1_output.sam exists\n",
      "sam file looks right\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing SAM file: 51it [00:00, 2973.24it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading results for bowtie2...\n",
      "File size: 0.02 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//bowtie2_output.sam exists\n",
      "sam file looks right\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing SAM file: 51it [00:00, 121126.56it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading results for minimap2...\n",
      "File size: 0.01 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//minimap2_output.sam exists\n",
      "sam file looks right\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing SAM file: 23it [00:00, 87778.88it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading results for bbmap_skimmer...\n",
      "File size: 0.01 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//bbmap_skimmer_output.sam exists\n",
      "sam file looks right\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing SAM file: 51it [00:00, 125313.12it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading results for strobealign...\n",
      "File size: 0.01 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//strobealign_output.sam exists\n",
      "sam file looks right\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing SAM file: 59it [00:00, 119030.27it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading results for blastn...\n",
      "File size: 0.00 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//blastn_output.tsv exists\n",
      "\n",
      "Reading results for mmseqs2...\n",
      "File size: 0.00 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//mmseqs_output.tsv exists\n",
      "qend < qstart detected, assuming tend and tstart are reversed\n",
      "\n",
      "Reading results for spacer_containment...\n",
      "File size: 0.00 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//spacer_containment_output.tsv exists\n",
      "\n",
      "Reading results for mummer4...\n",
      "File size: 0.01 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//mummer4_output.sam exists\n",
      "@HD\tVN:1.4\tSO:unsorted\n",
      "No reference file provided, using the smallest sam file present in the output file directory\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing SAM file: 51it [00:00, 108804.43it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading results for lexicmap...\n",
      "File size: 0.01 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//lexicmap_output.tsv exists\n",
      "\n",
      "Reading results for vsearch...\n",
      "File size: 0.00 MB\n",
      " output file  results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/raw_outputs//vsearch_output.blast6 exists\n",
      "shape: (11, 7)\n",
      "┌────────────────┬───────────────┬───────────────┬───────────────┬───────────┬──────────┬──────────┐\n",
      "│ tool           ┆ true_positive ┆ false_positiv ┆ ground_truth_ ┆ precision ┆ recall   ┆ f1_score │\n",
      "│ ---            ┆ s             ┆ es            ┆ total         ┆ ---       ┆ ---      ┆ ---      │\n",
      "│ str            ┆ ---           ┆ ---           ┆ ---           ┆ f64       ┆ f64      ┆ f64      │\n",
      "│                ┆ i64           ┆ i64           ┆ i64           ┆           ┆          ┆          │\n",
      "╞════════════════╪═══════════════╪═══════════════╪═══════════════╪═══════════╪══════════╪══════════╡\n",
      "│ bowtie1        ┆ 51            ┆ 0             ┆ 51            ┆ 1.0       ┆ 1.0      ┆ 1.0      │\n",
      "│ bowtie2        ┆ 51            ┆ 0             ┆ 51            ┆ 1.0       ┆ 1.0      ┆ 1.0      │\n",
      "│ minimap2       ┆ 17            ┆ 0             ┆ 51            ┆ 1.0       ┆ 0.333333 ┆ 0.5      │\n",
      "│ bbmap_skimmer  ┆ 51            ┆ 0             ┆ 51            ┆ 1.0       ┆ 1.0      ┆ 1.0      │\n",
      "│ …              ┆ …             ┆ …             ┆ …             ┆ …         ┆ …        ┆ …        │\n",
      "│ spacer_contain ┆ 51            ┆ 0             ┆ 51            ┆ 1.0       ┆ 1.0      ┆ 1.0      │\n",
      "│ ment           ┆               ┆               ┆               ┆           ┆          ┆          │\n",
      "│ mummer4        ┆ 51            ┆ 0             ┆ 51            ┆ 1.0       ┆ 1.0      ┆ 1.0      │\n",
      "│ lexicmap       ┆ 33            ┆ 0             ┆ 51            ┆ 1.0       ┆ 0.647059 ┆ 0.785714 │\n",
      "│ vsearch        ┆ 51            ┆ 0             ┆ 51            ┆ 1.0       ┆ 1.0      ┆ 1.0      │\n",
      "└────────────────┴───────────────┴───────────────┴───────────────┴───────────┴──────────┴──────────┘\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "python bench.py -t 14 -nc 400 -ns 10 -lm 0 0 -ir 1 10 -prc 0.5 -mr 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we'll modify the bash scripts to use the real data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/real_data/bash_scripts/bbmap_skimmer.sh\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/real_data/bash_scripts/blastn.sh\n",
      "results/real_data/bash_scripts/bowtie1.sh\n",
      "results/real_data/bash_scripts/bowtie2.sh\n",
      "results/real_data/bash_scripts/lexicmap.sh\n",
      "results/real_data/bash_scripts/minimap2.sh\n",
      "results/real_data/bash_scripts/mmseqs.sh\n",
      "results/real_data/bash_scripts/mummer4.sh\n",
      "results/real_data/bash_scripts/spacer_containment.sh\n",
      "results/real_data/bash_scripts/strobealign.sh\n",
      "results/real_data/bash_scripts/vsearch.sh\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /clusterfs/jgi/scratch/science/metagen/neri/code/blits/spacer_matching_bench\n",
    "mkdir -p results/real_data/\n",
    "mkdir -p results/real_data/slurm_logs/ results/real_data/results/ results/real_data/bash_scripts/ results/real_data/raw_outputs/ results/real_data/simulated_data/\n",
    "\n",
    "for script in $(ls results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5/bash_scripts/*.sh); do\n",
    "    new_script=results/real_data/bash_scripts/$(basename $script)\n",
    "    cp $script results/real_data/bash_scripts/\n",
    "    echo $new_script\n",
    "    sed -i '1s|^|#!/bin/bash\\n|' $new_script\n",
    "    sed -i '2s|^|cd /clusterfs/jgi/scratch/science/metagen/neri/code/blits/spacer_matching_bench/\\n|' $new_script\n",
    "    \n",
    "    # Replace all variations of the run_t path\n",
    "    sed -i 's|results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5|results/real_data|g' $new_script\n",
    "    sed -i 's|./results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5|results/real_data|g' $new_script\n",
    "    sed -i 's|/results/run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5|/results/real_data|g' $new_script\n",
    "    sed -i 's|run_t_14_nc_400_ns_10_ir_1_10_lm_0_0_prc_0.5|real_data|g' $new_script\n",
    "    \n",
    "    # Replace data paths\n",
    "    sed -i 's|results/real_data/simulated_data/simulated_contigs.fa|imgvr4_data/crispr-host-prediction/IMGVR_all_nucleotides-high_confidence_nr.fna|g' $new_script\n",
    "    sed -i 's|results/real_data/simulated_data/simulated_spacers.fa|imgvr4_data/crispr-host-prediction/crispr-db/rm_duped_spacers.fna|g' $new_script\n",
    "    \n",
    "    # Remove hyperfine wrapper\n",
    "    sed -i 's/hyperfine.*json '\\''\\(.*\\)'\\''/\\1/' $new_script\n",
    "    \n",
    "    # Change thread count\n",
    "    sed -i 's|14|\"$1\"|g' $new_script\n",
    "    \n",
    "    # Split && into multiple lines\n",
    "    sed -i 's/ && /\\n/g' $new_script\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we'll make a slurm scripts to submit the tools as seperate jobs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p results/real_data/job_scripts/\n",
    "for script in $(ls results/real_data/bash_scripts/*.sh); do\n",
    "    tool_name=$(basename $script .sh)\n",
    "    script_full_path=$(realpath $script)\n",
    "    \n",
    "    echo \"#!/bin/bash\n",
    "#SBATCH --mail-user=uneri@lbl.gov\n",
    "#SBATCH --mail-type=FAIL,END,BEGIN\n",
    "#SBATCH -A grp-org-sc-metagen\n",
    "#SBATCH -q jgi_normal\n",
    "#SBATCH -c \n",
    "#SBATCH --mem=168G \n",
    "## specify runtime\n",
    "#SBATCH -t 72:00:00\n",
    "## specify job name\n",
    "#SBATCH -J \"$tool_name\"   \n",
    "## specify output and error file\n",
    "#SBATCH -o /clusterfs/jgi/scratch/science/metagen/neri/code/blits/spacer_matching_bench/results/real_data/slurm_logs/\"$tool_name\"-%A.out\n",
    "#SBATCH -e /clusterfs/jgi/scratch/science/metagen/neri/code/blits/spacer_matching_bench/results/real_data/slurm_logs/\"$tool_name\"-%A.err\n",
    "bash $script_full_path \"\\$\"SLURM_CPUS_PER_TASK\n",
    "\" > results/real_data/job_scripts/\"$tool_name\".sh\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and run the tools as seperate jobs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/real_data/job_scripts/bbmap_skimmer.sh\n",
      "results/real_data/job_scripts/blastn.sh\n",
      "results/real_data/job_scripts/bowtie1.sh\n",
      "results/real_data/job_scripts/bowtie2.sh\n",
      "results/real_data/job_scripts/lexicmap.sh\n",
      "results/real_data/job_scripts/minimap2.sh\n",
      "results/real_data/job_scripts/mmseqs.sh\n",
      "results/real_data/job_scripts/mummer4.sh\n",
      "results/real_data/job_scripts/spacer_containment.sh\n",
      "results/real_data/job_scripts/strobealign.sh\n",
      "results/real_data/job_scripts/vsearch.sh\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for script in $(ls results/real_data/job_scripts/*.sh); do\n",
    "    echo $script\n",
    "    sbatch $script\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next (in [`Performance_imgvr4.ipynb`](./Performance_imgvr4.ipynb)), we'll use functions from bench.py to read the mapping/search results and compare them."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
